contrib/babelfishpg_tsql/src/hooks.c

```
@@ -5623,6 +5632,204 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+ * Function : handle_group_by_percent_count(SelectStmt *groupbySelectStmt) 
+ *
+ * For regular queries, uses simple COUNT(*) from the table.
+ */
+static void 
+transform_percent_clause(ParseState *pstate, SelectStmt *stmt)
+{
+    /* To create inner select count(*) from t for percent */
+    SelectStmt    *totalCountSelectStmt; 
+    
+    /* Variables for building percentage calculation expression */
+    SubLink       *countSublink;     /* Subquery link for wrap select COUNT(*).. */
+    A_Expr        *mulExpr;          /* to multiply N*(select count(*)..) */
+    A_Expr        *divExpr;          /* Division expression to divide by 100 */
+    A_Const       *intConst;         /* Integer constant (100) */
+    FuncCall      *ceilFunc;         /* CEIL function for rounding */
+    
+    /* Check whether TSQL and limit value present */
+	if (sql_dialect != SQL_DIALECT_TSQL || stmt->limitCount == NULL)
+        return;
+    
+    if(stmt->limitOption != LIMIT_OPTION_PERCENT)
+        return;
+    else
+        stmt->limitOption = LIMIT_OPTION_COUNT;
+
+    /* Add validation check for percentage value > 100 */
+    if (IsA(stmt->limitCount, A_Const))
+    {
+        A_Const *const_val = (A_Const *) stmt->limitCount;
+        if (const_val->val.ival.type == T_Integer)
+        {
+            int percent_val = const_val->val.ival.ival;
+            if (percent_val > 100)
+                ereport(ERROR,
+                        (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+                        errmsg("Percent values must be between 0 and 100.")));
+        }
+        else if (const_val->val.fval.type == T_Float)
+        {
+            float8 percent_val = strtod(const_val->val.fval.fval, NULL);
+            if (percent_val > 100.0)
+                ereport(ERROR,
+                        (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+                        errmsg("Percent values must be between 0 and 100.")));
+        }
```

What about else condition ? If this is never expected to happen lets throw an error.

---

+1

---

```
if percentage > 100 
    then we wanted to throw error message.
else 
   It should run transform percent clause
```
the transformation logic is basically a else condition